### PR TITLE
Added ProtectionManager

### PR DESCRIPTION
**Pull request by StefanCostea**
- Implements Protection class that is used to watch over values of variables and implement protections over it, that will trigger whenever the variable don't fullfill its protection expected values

- Implements a ProtectionManager that is a container class for a group of protections and will check if any of the protections triggers and perform actions and notify it whenever happens.

ProtectionHandler can monitor and notify the following variables: name of the variable, value of the variable when triggered, value of the protection expected value, type of the protection and the board that contained the protection.

The ProtectionManager is used as follows:

```cpp
#define BOARD_ID 54
ProtectionManager::set_id(BOARD_ID);

double variable = 5.3;
ProtectionManager_add_protection(&variable, OUT_OF_RANGE,1,20,BELOW,5,EQUALS,5.2);
ProtectionManager::check_protections();
```

The protection will trigger whenever variable gets below 1 or above 20 or below 5 or when it is equal to 5.2.

### CODE REVIEW

File: Inc/ST-LIB_HIGH/Protections/Protection.hpp [hunk 1]
```diff
@@ -0,0 +1,40 @@
+#pragma once
+
+#include "C++Utilities/CppUtils.hpp"
+
+#define PROTECTIONTYPE_LENGTH 5
+
+enum ProtectionType : uint64_t {
+    BELOW = std::numeric_limits<uint64_t>::max() - PROTECTIONTYPE_LENGTH + 1,
```
Comment (PabloGonzalezSegarra) [inline Inc/ST-LIB_HIGH/Protections/Protection.hpp:8]: Porque no?
```suggestion
    BELOW = UINT64_MAX - PROTECTIONTYPE_LENGTH + 1,
```
Es una macro que especifica la HAL.

File: Inc/ST-LIB_HIGH/Protections/ProtectionManager.hpp [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+    static void set_id(int board_id);
+
+    template<class... AdditionalProtections>
+    static Protection& _add_protection(double* src, ProtectionType protection_type,double bound, AdditionalProtections... more_protections);
+    template<class... AdditionalProtections>
+    static Protection& _add_protection(double* src, ProtectionType protection_type, double lower_bound, double upper_bound, AdditionalProtections... more_protections);
+
+    static void check_protections();
+};
```
Comment (PabloGonzalezSegarra) [inline Inc/ST-LIB_HIGH/Protections/ProtectionManager.hpp:24]: Te falta una linea alfinal.

File: Src/ST-LIB_HIGH/Protections/Protection.cpp [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+            }
+            break;
+        case NOT_EQUALS:
+            if (*src != boundaries[boundary_counter++]) {
+                jumped_potection = protection;
+                return false;
+            }
+            break;
+        default:
+            break;
```
Comment (PabloGonzalezSegarra) [inline Src/ST-LIB_HIGH/Protections/Protection.cpp:82]: Pon `TODO: Error handler` aunque sea imposible que se meta en default.

File: Src/ST-LIB_HIGH/Protections/Protection.cpp [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        }
+
+    }
+    boundary_counter = 0;
+    return true;
+}
+
+map<ProtectionType, string> Protection::protection_type_name = { {ProtectionType::BELOW,"minimo"},
```
Comment (PabloGonzalezSegarra) [inline Src/ST-LIB_HIGH/Protections/Protection.cpp:90]: Esto alfinal se usa?

File: Src/ST-LIB_HIGH/Protections/ProtectionManager.cpp [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+}
+
+void ProtectionManager::check_protections() {
+    for (Protection protection: protections) {
+        if (!protection.check_state()) {
+        	//TODO: Send StateMachine to FAULT state
+            string range = "";
+            if (protection.jumped_potection == ProtectionType::OUT_OF_RANGE) {
+                range = "[";
+                range += (* protection.src < protection.boundaries[protection.boundary_counter - 1]) ? to_string(protection.boundaries[protection.boundary_counter - 1]) + "," + to_string(protection.boundaries[protection.boundary_counter]) + "]" :
```
Comment (PabloGonzalezSegarra) [inline Src/ST-LIB_HIGH/Protections/ProtectionManager.cpp:26]: Cuando alejandro vea esta rama le da un derrama xD.

**Review by PabloGonzalezSegarra — COMMENTED** [review]
Mira lo que te he puesto.

File: Inc/ST-LIB_HIGH/Protections/Protection.hpp [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+class Protection {
+public:
+
+    double* src;
+    char* name = nullptr;
+    vector<ProtectionType> protections;
+    vector<double> boundaries;
+    uint8_t boundary_counter = 0;
+    ProtectionType jumped_potection = (ProtectionType)0;
```
Comment (alejandrogonzalvo) [inline Inc/ST-LIB_HIGH/Protections/Protection.hpp:24]:
```suggestion
    ProtectionType jumped_protection = (ProtectionType)0;
```
Añade la r a la variable

File: Src/ST-LIB_HIGH/Protections/ProtectionManager.cpp [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+                to_string(protection.boundaries[protection.boundary_counter-2]) + "," + to_string(protection.boundaries[protection.boundary_counter-1]) + "]";
+            }
+            else {
+                range = to_string(protection.boundaries[protection.boundary_counter - 1]);
+            }
+            //TODO: Send Packet through TCP with information
+        }
+    }
```
Comment (alejandrogonzalvo) [inline Src/ST-LIB_HIGH/Protections/ProtectionManager.cpp:34]:
```suggestion
    for (Protection protection: protections) {
        if (protection.check_state()) {
            //TODO: Send Packet through TCP with information
            continue;
        }

        //TODO: Send StateMachine to FAULT state
        string range = "";
        string penultimate_protection = to_string(protection.boundaries[protection.boundary_counter - 1]);
        if (protection.jumped_protection != ProtectionType::OUT_OF_RANGE) {
            range = penultimate_protection;
            //TODO: Send Packet through TCP with information
            continue;
        }

        ProtectionType& last_protection = to_string(protection.boundaries[protection.boundary_counter]);
        ProtectionType& antepenultimate_protection = to_string(protection.boundaries[protection.boundary_counter - 2]);
        if (*protection.src < penultimate_protection) {
            range = "[" + penultimate_protection + "," + last_protection + "]";
        }
        else {
            range = "[" + antepenultimate_protection + "," + penultimate_protection + "]";
        }

        //TODO: Send Packet through TCP with information
    }
```
Asi mucho mas legible, cambia penultimate_protection por un nombre mas apropiado lower_bound, o lo que haga en el contexto de la funcion

File: Src/ST-LIB_HIGH/Protections/Protection.cpp (hunk 3 above)
Comment (alejandrogonzalvo) [reply Src/ST-LIB_HIGH/Protections/Protection.cpp:82]: Y no hace falta un switch, todos los cases hacen lo mismo!

**Review by alejandrogonzalvo — CHANGES_REQUESTED** [review]
la mayoria bastante clean, refactoriza lo que te he puesto

**Pull request merged**